src/topology/ready_arrays.rs

```diff
@@ -0,0 +1,86 @@
+use std::pin::Pin;
+
+use futures::{
+    task::{Context, Poll},
+    {Stream, StreamExt},
+};
+
+use crate::event::{EventArray, EventContainer};
+
+const DEFAULT_CAPACITY: usize = 4096;
+
+/// A stream combinator aimed at improving the performance of event streams under load.
+///
+/// This is similar in spirit to `StreamExt::ready_chunks`, but built specifically `EventArray`.
+/// The more general `FoldReady` is left as an exercise to the reader.
+pub struct ReadyArrays<T> {
+    inner: T,
+    enqueued: Vec<EventArray>,
+    enqueued_size: usize,
+    enqueued_limit: usize,
+}
+
+impl<T> ReadyArrays<T>
+where
+    T: Stream<Item = EventArray> + Unpin,
+{
+    /// Create a new `ReadyArrays` by wrapping an event array stream.
+    pub fn new(inner: T) -> Self {
+        Self::with_capacity(inner, DEFAULT_CAPACITY)
```

Comment:
Is this really what you want to be doing here? This provides a pre-allocated capacity of 4096 *arrays*, which is a bit excessive.

Reply:
I dropped the default size and clarified the documentation in [17b52be](https://github.com/vectordotdev/vector/pull/12626/commits/17b52bebab9576aa0f4d0006677d8ca673709e08).

Reply:
This was all thrown together pretty hastily as a test, so certainly not ideal in many ways 😄 

That being said, there is at least some argument for a large-ish capacity here. Without it, we're limited in the case where a source is returning single-event arrays (or similar), or when some filtering/branching/sampling transform breaks up larger arrays into much smaller ones as they flow through the system. I'm still not sure that makes it worthwhile, but something to consider.

Reply:
Follow up: https://github.com/vectordotdev/vector/issues/12662